sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -135,154 +155,216 @@ namespace Azure { namespace Core { namespace Http {
+    void SetPort(uint16_t port) { m_port = port; }
+
+    void SetPath(const std::string& path, bool isPathEncoded = false)
+    {
+      m_path = isPathEncoded ? path : EncodePath(path);
+    }
+
+    void SetFragment(const std::string& fragment, bool isFragmentEncoded = false)
+    {
+      m_fragment = isFragmentEncoded ? fragment : EncodeFragment(fragment);
+    }
+
+    /******** API for mutating URL state ********/
+
+    // Path is mostly expected to be appended without url-encoding. Be default, path will be encoded
+    // before it is added to URL. \p isPathEncoded can set to true to avoid encoding.
+    void AppendPath(const std::string& path, bool isPathEncoded = false)
     {
-      // Constructor makes sure path never keeps slash at the end so we can feel OK on adding slash
-      // on every append path
-      this->m_path += "/" + path;
+      if (!m_path.empty() && m_path.back() != '/')
+      {
+        m_path += '/';
+      }
+      m_path += isPathEncoded ? path : EncodePath(path);
     }
-    std::string ToString() const
+
+    // the value from query parameter is mostly expected to be non-url-encoded and it will be
+    // encoded before adding to url by default. Use \p isValueEncoded = true when the value is
+    // already encoded.
+    //
+    // Note: a query key can't contain any chars that needs to be url-encoded. (by RFC).
+    //
+    // Note: AppendQuery override previous query parameters.
+    void AppendQuery(const std::string& key, const std::string& value, bool isValueEncoded = false)
     {
-      auto port = this->m_port.size() > 0 ? ":" + this->m_port : "";
-      return this->m_scheme + "://" + this->m_host + port + this->m_path;
+      if (m_retryModeEnabled)
+      {
+        m_retryQueryParameters[key] = value;
+      }
+      else
+      {
+        m_queryParameters[key] = value;
+      }
+
+      if (!isValueEncoded)
+      {
+        // Inserting a query that needs to be encoded
+        m_needToEncodeQueries.insert(key);
+      }
+      else
+      {
+        // Inserting encoded value. Make sure it is not in need encoding index
+        m_needToEncodeQueries.erase(key);
+      }
     }
-    std::string GetPath() const { return this->m_path; }
-    std::string GetHost() const { return this->m_host; }
-    std::map<std::string, std::string> GetQueryParameters() const
+
+    // removes a query parameter
+    void RemoveQuery(const std::string& key)
     {
-      return this->m_queryParameters;
+      m_queryParameters.erase(key);
+      m_retryQueryParameters.erase(key);
+      m_needToEncodeQueries.erase(key);
     }
-    void AddQueryParameter(std::string const& name, std::string const& value)
+
+    /************** API to read values from URL ***************/
+
+    std::string GetHost() const { return m_host; }
+
+    const std::string& GetPath() const { return m_path; }
+
+    // Copy from query parameters list. Query parameters from retry map have preference and will
```

Comment:
One thing I learn from C is that our docs/methods will keep changing frequently before GA and keeping the doxygen comments up to date is painful. Can we wait on it until we are close to GA or some time where we can just update to doxygen style?
Usually the problem is to keep parameters and return information in Doxygen up to date

Reply:
That is going to be a lot of accrued tech debt for late in the development cycle. I suggest going forward all the "new" docs be in the right/expected doxygen format. Renames are a lot easier and quick to fix and update compared to re-writing/re-formatting all the docs.

Reply:
I think this is not a big deal for now.
It is easier for me as I mentioned because it is a copy/paste using doxygen plugin from vscode to auto generate the formatting for it.

Once we have review the surface with Jeff, when we know the right names for the args and all those details, it is just copy and paste the info. The plugin takes args and everything and creates the right formatting

Reply:
Per Wednesday's meeting, doxygen comments/updates are required on all PRs moving forward.  This PR however was in the exemption bucket as it was an already created PR.